compiler/rustc_middle/src/mir/interpret/value.rs

```diff
@@ -41,12 +41,20 @@ pub enum ConstValue<'tcx> {
     /// Only for ZSTs.
     ZeroSized,
 
-    /// Used for `&[u8]` and `&str`.
+    /// Used for references to unsized types with slice tail.
     ///
-    /// This is worth an optimized representation since Rust has literals of these types.
-    /// Not having to indirect those through an `AllocId` (or two, if we used `Indirect`) has shown
-    /// measurable performance improvements on stress tests.
-    Slice { data: ConstAllocation<'tcx>, start: usize, end: usize },
+    /// This is worth an optimized representation since Rust has literals of type `&str` and
+    /// `&[u8]`. Not having to indirect those through an `AllocId` (or two, if we used `Indirect`)
+    /// has shown measurable performance improvements on stress tests. We then reuse this
+    /// optimization for slice-tail types more generally during valtree-to-constval conversion.
+    Slice {
+        /// The allocation storing the slice contents.
+        /// This always points to the beginning of the allocation.
+        data: ConstAllocation<'tcx>,
+        /// The metadata field of the reference.
+        /// This is a "target usize", so we use `u64` as in the interpreter.
+        meta: u64,
```

Comment:
do we even need this anymore? It can be computed from the `ConstAllocation`'s size and the element type from the constant's type

Reply:
This variant is only interesting for MIR building in check builds. For codegen builds we'll end-up creating the AllocId anyway.
Meawhile, we still have an asymmetry between the variants of `ConstValue` and the variants we have in interpreter and codegen (Uninit, Scalar, ScalarPair, Indirect).

What about:
- making this a variant in `mir::ConstantKind` instead? a `ConstantKind::ByteSlice(&'tcx [u8])`?
- introducing a `ScalarPair` variant in `ConstValue`? Example in https://github.com/rust-lang/rust/pull/115915 for the latter.

The "initial" representation by MIR building would be a `ConstantKind::ByteSlice(&[u8])`, and would get normalized to a `ScalarPair` for/before codegen.

Reply:
> It can be computed from the ConstAllocation's size and the element type from the constant's type

Not for ZST slices...

Reply:
> a ConstantKind::ByteSlice(&'tcx [u8])?

How does that make any sense? All fully evaluated constants should be `ConstantKind::Value`. That's literally what the variant is for. We absolutely should not have other variants carry evaluated constants.

Reply:
> Meawhile, we still have an asymmetry between the variants of ConstValue and the variants we have in interpreter and codegen (Uninit, Scalar, ScalarPair, Indirect).

Yeah that is totally deliberate. We used to have `ConstValue::ScalarPair` and moved away from it. I think moving towards ScalarPair would be a mistake; ScalarPair already have too much freedom in their layout.

I think long-term we want to remove `ConstValue` entirely. See https://github.com/rust-lang/rust/issues/115877.